R/differentiate-expr.R

```diff
@@ -0,0 +1,291 @@
+    maths$divide(differentiate(expr[[2]], name),
+                 maths$times(2, maths$rewrite(expr)))
+  },
+  `if` = function(expr, name) {
+    condition <- maths$rewrite(expr[[2]])
+    da <- differentiate(expr[[3]], name)
+    db <- differentiate(expr[[4]], name)
+    if (identical(da, db)) da else call("if", condition, da, db)
+  },
+  lfactorial = function(expr, name) {
+    a <- maths$rewrite(expr[[2]])
+    maths$times(differentiate(a, name), call("digamma", maths$plus(a, 1)))
+  },
+  abs = function(expr, name) {
+    a <- maths$rewrite(expr[[2]])
+    maths$times(differentiate(a, name), call("sign", a))
+  }
+)
+
+maths <- local({
+  .protect <- function(x, except) {
+    if (is.recursive(x)) {
+      fn <- as.character(x[[1]])
+      if (fn == "(") {
+        return(.protect(x[[2]], except))
+      }
+      pass <- grepl("^[a-z]", fn) ||
+        (length(except) > 0 && fn %in% except) ||
+        "unary_minus" %in% except && .is_unary_minus(x)
+      if (pass) {
+        return(x)
+      }
+      call("(", x)
+    } else {
+      x
+    }
+  }
+  .drop_parens <- function(x) {
+    if (is_call(x, "(")) .drop_parens(x[[2]]) else x
+  }
+  .is_unary_minus <- function(expr, recurse = FALSE) {
+    (is.numeric(expr) && expr < 0) ||
+      (is_call(expr, "-") && length(expr) == 2) ||
+      (recurse && (
+        (is_call(expr, "*") || is_call(expr, "/")) &&
+        .is_unary_minus(expr[[2]], TRUE)))
+  }
+  is_zero <- function(x) {
+    is.numeric(x) && x == 0
+  }
+  is_one <- function(x) {
+    is.numeric(x) && x == 1
+  }
+  is_minus_one <- function(x) {
+    is.numeric(x) && x == -1
+  }
+  plus <- function(a, b) {
+    ## there's more cancelling here that could be done with
+    ## expressions that involve subtraction and numbers, but I don't
+    ## see them turning up anywhere yet; probably best to implement
+    ## after we have them.
+    if (is.numeric(a) && is.numeric(b)) {
+      a + b
+    } else if (is.numeric(b)) {
+      plus(b, a)
+    } else if (is_zero(a)) {
+      .drop_parens(b)
+    } else if (is_call(b, "+")) {
+      plus(plus(a, b[[2]]), b[[3]])
+    } else {
+      call("+", a, b)
+    }
+  }
+  minus <- function(a, b) {
+    if (is.numeric(a) && is.numeric(b)) {
+      a - b
+    } else if (is_zero(b)) {
+      .drop_parens(a)
+    } else if (is_zero(a)) {
+      uminus(b)
+    } else {
+      call("-", a, .protect(b, c("*", "/", "^")))
+    }
+  }
+  uminus <- function(a) {
+    if (is.numeric(a)) {
+      -a
+    } else if (.is_unary_minus(a)) {
+      a[[2]]
+    } else if (is_call(a, "*")) {
+      if (.is_unary_minus(a[[2]])) {
+        times(a[[2]][[2]], a[[3]])
+      } else if (.is_unary_minus(a[[3]])) {
+        times(a[[2]], a[[3]][[2]])
+      } else {
+        times(uminus(a[[2]]), a[[3]])
+      }
+    } else if (is_call(a, "/")) {
+      maths$divide(maths$uminus(a[[2]]), a[[3]])
+    } else if (is_call(a, "-") && length(a) == 3) {
+      minus(a[[3]], a[[2]])
+    } else if (is_call(a, "(")) {
```

Comment:
just felt natural to add this one in since you are already kind of distributing -(a - b) -> b - a so can also do -(a + b) -> -a - b
```suggestion
    } else if (is_call(a, "-") && length(a) == 3) {
      minus(a[[3]], a[[2]])
    } else if (is_call(a, "+")) {
      minus(uminus(a[[2]]), a[[3]])
    } else if (is_call(a, "(")) {
```

Reply:
I think I prefer `-(a + b)` here tbh